File: src/solution/s0001_two_sum.rs

use std::collections::HashMap;

pub struct Solution {}

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut nums_map = HashMap::new();
        
        for (index, num) in nums.iter().enumerate(){
            let complement = target - num; // subtract to avoid overflow

            if let Some(sub_index) = nums_map.get(&complement){
                return vec![*sub_index, index as i32]
            }

            nums_map.insert(num, index as i32);
        }

        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(vec![0, 1], Solution::two_sum(vec![2, 7, 11, 15], 9));
        assert_eq!(vec![1, 2], Solution::two_sum(vec![3, 2, 4], 6));
    }
}

File: python/s0001_two_sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_sorted = sorted(nums)
        i, j = 0, len(nums)-1 
        # 1 2 3 4 5 6 7 8

        while i < j:
            if nums_sorted[i] + nums_sorted[j] > target:
                j -= 1
            elif nums_sorted[i] + nums_sorted[j] < target:
                i += 1
            else:
                break

        res = []
        for index in range(len(nums)):
            if nums[index] == nums_sorted[i] or nums[index] == nums_sorted[j]:
                res.append(index)

        return res